### PR TITLE
Log all rejected requests (#1295)

### PR DESCRIPTION
**Pull request by cesarbs**
#1295 is caused by there being places in the code where `BadHttpRequestException`s are thrown but not logged.

This PR changes the places where `BadHttpRequestException` is thrown (other than `Frame.RejectRequest()`) to throw a new exception type (`DecodingException`). This fixes layering since those places should not be throwing `BadHttpRequestException` (gross to do so in a place like `MemoryPoolIterator.GetAsciiString()`, for example). The new exception type is caught where required and `Frame.RejectRequest()` is called, effectively converting the lower-level exception into a `BadHttpRequestException`.

Additionally, I've done significant refactoring in `BadHttpRequestTests`.

I'll send another PR shortly where I fixed the same issue but in a different way. Still think this is the right one though.

Need to consider switching the place where we log the bad request from `Frame.RejectRequest()` to wherever it's caught (two places in `Frame<T>`). That would avoid having new code that throws it not via `Frame.RejectRequest()` getting bad requests not logged. Appreciate thoughts on this.

@halter73 @natemcmaster @Tratcher @pakrym

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 1]
```diff
@@ -1180,6 +1180,13 @@ public RequestLineStatus TakeStartLine(SocketInput input)
 
                 return RequestLineStatus.Done;
             }
+            catch (DecodingException)
+            {
+                RejectRequest(RequestRejectionReason.NonAsciiOrNullCharactersInRequestLine,
+                    Log.IsEnabled(LogLevel.Information) ? start.GetAsciiStringEscaped(end, MaxInvalidRequestLineChars) : string.Empty);
+                // won't run but compiler complains
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Would it work if we changed the return type of 'RejectRequest*' methods to Exception and instead did `throw RejectRequest(...` ?

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Need to be careful to `throw` every time. I did consider this a while back. I prefer throwing from inside the method.

Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Fine either way, but for readability I actually prefer the explicit throw. The first time I read this code, it was not immediately obvious that 'RejectRequest' and its overloads terminate execution.

**Comment by halter73** [comment]
Nice catch and nice change. I want to get this in soon after #1277 is merged.

**Comment by cesarbs** [comment]
@davidfowl Keeping this open to merge after pipelines. Will probably require some re-work.

**Comment by davidfowl** [comment]
@cesarbs are you going to update this change and apply it to both branches?

**Comment by cesarbs** [comment]
Holding the change against `feature/dev-si` for now. This was done in `dev` in ac60f13312b0776cbffe333e72ce521211845b89 and 11c7eb5665b294395db4227dd6f163e7e0fcbed5.

**Pull request closed without merging**